once_test.py

```diff
@@ -360,6 +456,68 @@ def value():
         self.assertEqual(_CallOnceClass.value(), 1)
         self.assertEqual(_CallOnceClass.value(), 1)
 
+    def test_receiving_iterator(self):
+        @once.once
+        def receiving_iterator():
+            next = yield 0
+            while next is not None:
+                next = yield next
+
+        gen_1 = receiving_iterator()
+        gen_2 = receiving_iterator()
+        self.assertEqual(gen_1.send(None), 0)
+        self.assertEqual(gen_1.send(1), 1)
+        self.assertEqual(gen_1.send(2), 2)
+        self.assertEqual(gen_2.send(None), 0)
+        self.assertEqual(gen_2.send(-1), 1)
+        self.assertEqual(gen_2.send(-1), 2)
+        self.assertEqual(gen_2.send(5), 5)
+        self.assertEqual(next(gen_2, None), None)
+        self.assertEqual(gen_1.send(None), 5)
+        self.assertEqual(next(gen_1, None), None)
+        self.assertEqual(list(receiving_iterator()), [0, 1, 2, 5])
+
+    def test_receiving_iterator_parallel_execution(self):
+        @once.once
+        def receiving_iterator():
+            next = yield 0
+            while next is not None:
+                next = yield next
+
+        def call_iterator(_):
+            gen = receiving_iterator()
+            result = []
+            result.append(gen.send(None))
+            for i in range(1, 32):
+                result.append(gen.send(i))
+            return result
+
+        with concurrent.futures.ThreadPoolExecutor(max_workers=16) as executor:
+            results = list(executor.map(call_iterator, range(32)))
+        for result in results:
+            self.assertEqual(result, list(range(32)))
+
+    def test_receiving_iterator_parallel_execution_halting(self):
+        @once.once
+        def receiving_iterator():
+            next = yield 0
+            while next is not None:
+                next = yield next
+
+        def call_iterator(n):
+            """Call the iterator but end early"""
+            gen = receiving_iterator()
```

Comment:
What's the memory story for these generators that never get finished? When i was writing the async tests, I got warnings if I did that and had to make sure to close them all out.

Reply:
AFAIK, it's safe in sync, unlike async where you are expected to await all of your promises.